formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt

```diff
@@ -0,0 +1,202 @@
+    }
+
+    override fun equals(other: Any?): Boolean =
+        other is CborNegativeInt && other.value == value && other.tags.contentEquals(tags)
+
+    override fun hashCode(): Int = value.hashCode() * 31 + tags.contentHashCode()
+}
+
+/**
+ * Class representing unsigned CBOR integer (major type 0).
+ */
+@Serializable(with = CborUIntSerializer::class)
+public class CborPositiveInt(
+    public val value: ULong,
+    tags: ULongArray = ulongArrayOf()
+) : CborPrimitive(tags) {
+
+    override fun equals(other: Any?): Boolean =
+        other is CborPositiveInt && other.value == value && other.tags.contentEquals(tags)
+
+    override fun hashCode(): Int = value.hashCode() * 31 + tags.contentHashCode()
+}
+
+/**
+ * Class representing CBOR floating point value (major type 7).
+ */
+@Serializable(with = CborDoubleSerializer::class)
+public class CborDouble(
+    public val value: Double,
+    tags: ULongArray = ulongArrayOf()
+) : CborPrimitive(tags) {
+
+    override fun equals(other: Any?): Boolean =
+        other is CborDouble && other.value == value && other.tags.contentEquals(tags)
+
+    override fun hashCode(): Int = value.hashCode() * 31 + tags.contentHashCode()
+}
+
+/**
+ * Class representing CBOR string value.
+ */
+@Serializable(with = CborStringSerializer::class)
+public class CborString(
+    public val value: String,
+    tags: ULongArray = ulongArrayOf()
+) : CborPrimitive(tags) {
+
+    override fun equals(other: Any?): Boolean =
+        other is CborString && other.value == value && other.tags.contentEquals(tags)
+
+    override fun hashCode(): Int = value.hashCode() * 31 + tags.contentHashCode()
+}
+
+/**
+ * Class representing CBOR boolean value.
+ */
+@Serializable(with = CborBooleanSerializer::class)
+public class CborBoolean(
+    private val value: Boolean,
+    tags: ULongArray = ulongArrayOf()
+) : CborPrimitive(tags) {
+
+    /**
+     * Returns the boolean value.
+     */
+    public val boolean: Boolean get() = value
+
+    override fun equals(other: Any?): Boolean =
+        other is CborBoolean && other.value == value && other.tags.contentEquals(tags)
+
+    override fun hashCode(): Int = value.hashCode() * 31 + tags.contentHashCode()
+}
+
+/**
+ * Class representing CBOR byte string value.
+ */
+@Serializable(with = CborByteStringSerializer::class)
+public class CborByteString(
+    private val value: ByteArray,
+    tags: ULongArray = ulongArrayOf()
+) : CborPrimitive(tags) {
+
+    /**
+     * Returns the byte array value.
+     */
+    public val bytes: ByteArray get() = value.copyOf()
+
+    override fun equals(other: Any?): Boolean =
+        other is CborByteString && other.value.contentEquals(value) && other.tags.contentEquals(tags)
+
+    override fun hashCode(): Int = value.contentHashCode() * 31 + tags.contentHashCode()
+}
+
+/**
+ * Class representing CBOR `null` value
+ */
+@Serializable(with = CborNullSerializer::class)
+public class CborNull(tags: ULongArray=ulongArrayOf()) : CborPrimitive(tags) {
```

Comment:
The default value should remain, but use an internal array (or null) to avoid extensive copies of empty arrays.